source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,173 @@
+
+Engine sets the pinning automatically when starting the VM as opposed to the manual pinning that is
+done by the user and is static.
+
+## Owner
+
+-   Name: Saif Abu Saleh (sabusale)
+-   Email: sabusale@redhat.com
+
+## Responsibilities
+
+VDSM:
+
+- Makes sure that dedicated VMs vCPUs stay exclusively pinned to the given pCPU
+
+Engine:
+- Holds cached available cpus based on host capabilities and pinned CPUs
+- Schedule VM to run on best satisfying host based on the above cached data structure
+
+
+
+## Backend
+
+1. Add the new policies to CPU Pinning Policy enum
+2. Save host capabilities CPU topology in vds_dynamic
+3. Hold hostToAvailableCpu data structure:
+    it will be on this form `Map<Integer,Map<Integer,Map<Integer,List<String>>>>`
+   
+    Host id -> Map of Socket id to Map of Core id to list of cpus
+    (assuming host id, core id, socket id is unique)
+
+    **Save/Update cached data structure:**    
+
+    Option 1:
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * Poll every X seconds all VMs and find which VM’s have cpu pinning and update the shared CPU pinning data structure based on VM pinned CPUs and host capabilities
+
+    Option 2:
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * On every start/stop/migrate/… VM update the shared CPU pinning data structure
+    with used pinning
+
+   **Init cached data structure (on engine start/stop):**
+   
+    Option1:
+    * Will be calculated after engine started after getting host caps and VMs cpu pinnings
+    
+    Option2:
+    * Create table in DB and serialize/deserialize it from DB
+
+4. Scheduling
+    1. Start VM:
+        - Scheduling phase
+            - Reads hostToAvailableCpu from step 3
+            - Filter hosts that can't match virtual CPU topology (pCpu pinning should
+              be close to the virtual topology, e.g: 2 vCPU on same vSocket should be placed on same pSocket)
+            - Score all hosts to find best matching host with enough pCPUs 
+        - Run phase
+            - Based on Step 3 -Save/Update cached data structure
+            - Update VM CurrentCpuPinning DB field with the mapping
+
+    2. Stop/Destroy VM:
+        - Clear VM dynamic CPU pinning field
+        - Based on Step 3 -Save/Update cached data structure
+    
+3. Send vcpupin with cpusets and policy name in VM metadata
+   1. Send the policy name, available policy names: { none, pin, dedicated, siblings, isolate-threads }.
+   2. Send the cpusets (it will be stored in VM_Dynamic)
+
+4. Migrate VDSM call
+   1. Calculate vCPU to pCPU mapping based on the scheduled host
+   2. Send cpusets additional parameter
+   
+5. setNumberOfCpus VDSM call
+   1. Send cpusets additional parameter
+
+
+## Database
+
+* Save CPU mappings in field CurrentCpuPinning field in vm_dynamic table
+* Save CPU topology from host capabilities into vds_statistics
```

Comment:
ah ok, so that's what you meant before about the stats cycles - if we're talking about capabilities then it should go to vds_dynamic instead